packages/core-api/src/apis/system/ApiProvider.tsx

```diff
@@ -35,16 +45,25 @@ type ApiProviderProps = {
   children: ReactNode;
 };
 
-const Context = createContext<ApiHolder | undefined>(undefined);
+type ApiContextType = VersionedValue<{ 1: ApiHolder }> | undefined;
+const ApiContext = getOrCreateGlobalSingleton('api-context', () =>
+  createContext<ApiContextType>(undefined),
+);
 
 export const ApiProvider = ({
   apis,
   children,
 }: PropsWithChildren<ApiProviderProps>) => {
-  const parentHolder = useContext(Context);
-  const holder = parentHolder ? new ApiAggregator(apis, parentHolder) : apis;
-
-  return <Context.Provider value={holder} children={children} />;
+  const parentHolder = useContext(ApiContext)?.atVersion(1);
+  const versionedValue = useMemo(
```

Comment:
@joaomilho added memos to this and AppProvider to avoid unnecessary rerenders, which we also need to add to the later utils